src/xtgeo/xyz_common/_xyz_data.py

```diff
@@ -0,0 +1,625 @@
+                f"Cannot set wlogtype as {attrtype}, not in "
+                f"{list(_AttrType.__members__)}"
+            )
+
+        self.ensure_consistency()
+
+    def get_attr_record(self, name: str):
+        """Get a record for a named attribute."""
+        return self._attr_records[name]
+
+    def set_attr_record(self, name: str, record: Optional[dict]) -> None:
+        """Set a record for a named log."""
+
+        if name not in self._attr_types:
+            raise ValueError(f"No such attr_name: {name}")
+
+        if record is None and self._attr_types[name] == _AttrType.DISC:
+            record = {}
+        elif record is None and self._attr_types[name] == _AttrType.CONT:
+            record = CONT_DEFAULT_RECORD
+
+        if self._attr_types[name] == _AttrType.CONT and isinstance(
+            record, (list, tuple)
+        ):
+            if len(record) == 2:
+                self._attr_records[name] = tuple(record)  # prefer as tuple
+        elif self._attr_types[name] == _AttrType.CONT and isinstance(record, dict):
+            raise ValueError(
+                "Cannot set a log record for a continuous log: input record is "
+                "dictionary, not a list or tuple"
+            )
+        elif self._attr_types[name] == _AttrType.DISC and isinstance(record, dict):
+            self._attr_records[name] = record
+        elif self._attr_types[name] == _AttrType.DISC and not isinstance(record, dict):
+            raise ValueError(
+                "Input is not a dictionary. Cannot set a log record for a discrete log"
+            )
+        else:
+            raise ValueError(
+                "Something went wrong when setting logrecord: "
+                f"({self._attr_types[name]} {type(record)})."
+            )
+
+        self.ensure_consistency()
+
+    def get_dataframe_copy(
+        self,
+        infer_dtype: bool = False,
+        filled=False,
+        fill_value=const.UNDEF_CONT,
+        fill_value_int=const.UNDEF_DISC,
+    ):
+        """Get a deep copy of the dataframe, with options.
+
+        If infer_dtype is True, then DISC columns will be of "int32" type, but
+        since int32 do not support np.nan, the value for undefined values will be
+        ``fill_value_int``
+        """
+        dfr = self._df.copy()
+        if infer_dtype:
+            for name, attrtype in self._attr_types.items():
+                if attrtype.name == "DISC":
+                    dfr[name] = dfr[name].fillna(fill_value_int)
+                    dfr[name] = dfr[name].astype("int32")
+
+        if filled:
+            dfill = {}
+            for attrname in self._df:
+                if self._attr_types[attrname] == _AttrType.DISC:
+                    dfill[attrname] = fill_value_int
+                else:
+                    dfill[attrname] = fill_value
+
+            dfr = dfr.fillna(dfill)
+
+        return dfr
+
+    def get_dataframe(self):
+        """Get the dataframe."""
+        return self._df
```

Comment:
Are both this (and `_set`) as well as the decorated `@property dataframe` needed?

Reply:
Perhaps not (not sure), so leave as-is for now. It is a non-public class anyway